lib/fs.js

```diff
@@ -0,0 +1,61 @@
+'use strict';
+
+const fs = require('fs');
+const path = require('path');
+const metasync = require('metasync');
+
+const rmdirp = (dir, cb) => {
+  fs.stat(dir, (err, stats) => {
+    if (err) {
+      if (err.code === 'ENOENT') cb(null);
+      else cb(err);
+      return;
+    }
+    if (stats.isDirectory()) {
+      fs.readdir(dir, (err, files) => {
+        if (err) {
+          cb(err);
+          return;
+        }
+        files = files.map(file => path.join(dir, file));
+        metasync.each(files, rmdirp, err => {
+          if (err) cb(err);
+          else fs.rmdir(dir, cb);
+        });
+      });
+    } else {
+      fs.unlink(dir, cb);
+    }
+  });
+};
+
+const mkdir = (dir, mode, cb) => {
```

Comment:
AFAIU rmdir and mkdirp use the other function for tests, because of this it is hard to make them truly independent.

Reply:
Yes, but function tests shouldn't depend on other function.  For  example, test for `rmdirp` can look something like this:
```js
metatests.test('rmdirp test', test =>
  fs.mkdir('./testDir', err =>
    fs.mkdir('./testDir/subdir1', err =>
      fs.mkdir('./testDir/subdir1/subdir2',
        rmdirp('./testDir', err =>
          fs.access('./testDir', err => {
            test.isError(err);
            test.end();
          })
        )
      )
    )
  )
);
```

Reply:
@SemenchenkoVitaliy It's not that simple: if a directory exists `mkdir` will result in an error. You should also check an error in case it is a real fs access error, which will transform your test into:
```js
metatests.test('rmdirp test', test =>
  fs.mkdir('./testDir', err => {
    if (err && err.code !== 'EEXISTS') {
      test.bailout(err);
    }
    fs.mkdir('./testDir/subdir1', err => {
      if (err && err.code !== 'EEXISTS') {
        test.bailout(err);
      }
      fs.mkdir('./testDir/subdir1/subdir2', err => {
        if (err && err.code !== 'EEXISTS') {
          test.bailout(err);
        }
        rmdirp('./testDir', err => {
          if (err && err.code !== 'EEXISTS') {
            test.bailout(err);
          }
          fs.access('./testDir', err => {
            test.isError(err);
            test.end();
          });
        });
      });
    });
  })
);

```

Reply:
Of course, there should be errors checks. I just proposed a concept code. As you can see, it doesn't even have error handling. Your example is something that should actually be used.

Reply:
And of course, we should make cleanup after test if it fails.

Reply:
> Your example is something that should actually be used.

I don't feel like it because there is a lot of boilerplate code it that case, and if we try to make some helper function we'll end up implementing both `mkdirp` and `rmdir`.

> And in case if the test fails, cleanup should be performed.

It's not always possible. The only way for preparation to fail is to have some issues with the filesystem, in that case cleanup is useless, because it will probably fail too.

Reply:
I think that if we have only 2 levels of nesting i.e. `./testDir` and `./testDir/subdir1` it won't look too bad. What I want to say is that tests for `mkdipr` shouldn't use `rmdirp` and vise versa. Yeah, sure, tests will look awful but I don't think we can do something about this. Moreover, `mkdirp` and `rmdipr` have their own PRs. Of course, we can create new PR, which will implement both `mkdirp` and `rmdirp` but IMO we should land those PRs for each function separately with their own dedicated tests and then fix tests or add more if it is needed.